File: Assets/Plugins/VoxelBusters/EssentialKit/Runtime/iOS/BillingServices/BillingProduct.cs
#if UNITY_IOS || UNITY_TVOS
using System;
using System.Collections.Generic;
using VoxelBusters.CoreLibrary.NativePlugins.iOS;

namespace VoxelBusters.EssentialKit.BillingServicesCore.iOS
{
    internal sealed class BillingProduct : BillingProductBase
    {
        #region Fields
        
        private     string      m_localizedTitle;

        private     string      m_localizedDescription;

        private     BillingPrice      m_price;

        private readonly List<BillingProductOffer> m_offers;

        private     BillingProductSubscriptionInfo m_subscriptionInfo;

        #endregion

        #region Constructors

        public BillingProduct(IntPtr nativeObjectPtr, string id, 
            string platformId, BillingProductType type, string localizedTitle, 
            string localizedDescription, BillingPrice price,
            BillingProductSubscriptionInfo subscriptionInfo,
            List<BillingProductOffer> offers,
            IEnumerable<BillingProductPayoutDefinition> payouts)
            : base(id: id, platformId: platformId, type: type, payouts: payouts)
        {
            // set properties
            NativeObjectRef         = new IosNativeObjectRef(nativeObjectPtr);
            m_localizedTitle        = localizedTitle;
            m_localizedDescription  = localizedDescription;
            m_price                 = price;
            m_offers                = offers;
            m_subscriptionInfo      = subscriptionInfo;
        }

        ~BillingProduct()
        {
            Dispose(false);
        }

        #endregion

        #region Base methods

        protected override string GetLocalizedTitleInternal()
        {
            return m_localizedTitle;
        }

        protected override string GetLocalizedDescriptionInternal()
        {
            return m_localizedDescription;
        }

        protected override BillingPrice GetPriceInternal()
        {
            return m_price;
        }

        protected override IEnumerable<BillingProductOffer> GetOffersInternal()
        {
            return m_offers;
        }

        protected override BillingProductSubscriptionInfo GetSubscriptionInfoInternal()
        {
            return m_subscriptionInfo;
        }


        #endregion
    }
}
#endif

File: Assets/Plugins/VoxelBusters/EssentialKit/Plugins/iOS/BillingServices/Modules/BillingServices/Types/Product/BillingProduct.swift
//
//  BillingProduct.swift
//  BillingServices
//
//  Created by Ayyappa on 26/03/24.
//

import Foundation

@objcMembers
public class BillingProduct : NSObject  {
    
    public private(set) var identifier: String
    public private(set) var localizedTitle: String
    public private(set) var localizedDescription: String
    public private(set) var price: BillingPrice
    public private(set) var subscriptionInfo: BillingProductSubscriptionInfo?
    public private(set) var offers: Array<BillingProductOffer>
    
    internal init(identifier: String, localizedTitle: String, localizedDescription: String, price: BillingPrice, subscriptionInfo: BillingProductSubscriptionInfo? = nil, offers: Array<BillingProductOffer>) {
        self.identifier = identifier
        self.localizedTitle = localizedTitle
        self.localizedDescription = localizedDescription
        self.price = price
        self.subscriptionInfo = subscriptionInfo
        self.offers = offers;
    }
}
